pkg/compact/compact.go

```diff
@@ -573,42 +575,80 @@ func (cg *Group) compact(ctx context.Context, dir string, comp tsdb.Compactor) (
 	for _, pdir := range plan {
 		meta, err := block.ReadMetaFile(pdir)
 		if err != nil {
-			return compID, errors.Wrapf(err, "read meta from %s", pdir)
+			return false, errors.Wrapf(err, "read meta from %s", pdir)
 		}
 
 		if cg.Key() != GroupKey(*meta) {
-			return compID, halt(errors.Wrapf(err, "compact planned compaction for mixed groups. group: %s, planned block's group: %s", cg.Key(), GroupKey(*meta)))
+			return false, halt(errors.Wrapf(err, "compact planned compaction for mixed groups. group: %s, planned block's group: %s", cg.Key(), GroupKey(*meta)))
 		}
 
 		for _, s := range meta.Compaction.Sources {
 			if _, ok := uniqueSources[s]; ok {
-				return compID, halt(errors.Errorf("overlapping sources detected for plan %v", plan))
+				return false, halt(errors.Errorf("overlapping sources detected for plan %v", plan))
 			}
 			uniqueSources[s] = struct{}{}
 		}
 
 		id, err := ulid.Parse(filepath.Base(pdir))
 		if err != nil {
-			return compID, errors.Wrapf(err, "plan dir %s", pdir)
+			return false, errors.Wrapf(err, "plan dir %s", pdir)
 		}
 
 		if err := block.Download(ctx, cg.bkt, id, pdir); err != nil {
-			return compID, retry(errors.Wrapf(err, "download block %s", id))
+			return false, retry(errors.Wrapf(err, "download block %s", id))
 		}
 
 		// Ensure all input blocks are valid.
-		if err := block.VerifyIndex(filepath.Join(pdir, block.IndexFilename), meta.MinTime, meta.MaxTime); err != nil {
-			return compID, halt(errors.Wrapf(err, "invalid plan block %s", pdir))
+		stats, err := block.GatherIndexIssueStats(filepath.Join(pdir, block.IndexFilename), meta.MinTime, meta.MaxTime)
+		if err != nil {
+			return false, errors.Wrapf(err, "gather index issues for block %s", pdir)
+		}
+
+		if err = stats.CriticalErr(); err == nil {
```

Comment:
You probably want to use `if err := ...` here, since you're in a different scope. Same for a few more `if` conditions below.

Reply:
also `err != nil`?

Reply:
ups